registered-comuni-pn/src/utils/helpers.utils.ts

```diff
@@ -0,0 +1,62 @@
+import { Purpose, PurposeState } from '../models/purposes.models.js'
+
+/**
+ * Converts an array of objects to CSV format.
+ * If the array is empty, an empty string is returned.
+ *
+ * @param data - The array of objects to convert to CSV
+ * @returns The CSV string
+ */
+export function toCSV(data: Array<Record<string, string>>) {
+  if (data.length === 0) return ''
+
+  const headers = Object.keys(data[0]).join(',') + '\n'
+  const columns = data.map((row) => Object.values(row).join(',')).join('\n')
+
+  return headers + columns
+}
+
+/**
+ * An extended Map that throws an error if the key is not found.
+ */
+export class SafeMap<K, V> extends Map<K, V> {
+  constructor(...args: ConstructorParameters<typeof Map<K, V>>) {
+    super(...args)
+  }
+
+  get(key: K): V {
+    const value = super.get(key)
+    if (!value) throw new Error(`Value not found for key: ${key}`)
+
+    return value
+  }
+}
+
+/**
+ * Transforms a purpose to a CSV output row.
+ */
+export function toCsvDataRow(tenantNamesMap: SafeMap<string, string>, purpose: Purpose) {
+  function getPurposeVersionWithState(state: PurposeState) {
+    return purpose.versions.find((version) => version.state === state)
+  }
+
+  const activeVersion = getPurposeVersionWithState('Active')
+  const suspendedVersion = getPurposeVersionWithState('Suspended')
+  const waitingForApprovalVersion = getPurposeVersionWithState('WaitingForApproval')
+
+  const relevantVersion = activeVersion || suspendedVersion || waitingForApprovalVersion
```

Comment:
Yes it is. The logic is the following:

- Is there at least one version with "Active"? The "stato_finalita_migliore" column should be "Attivo";
- If there is no active version, is there one with "Suspended" state? Than the column should be "Sospeso";
- If the purpose has no active/suspended version, but has a "WaitingForApproval" one, than the column will be "In attesa di attivazione";
- If no the purpose has version with neither of those state, the purpose should not be there. (This should never be the case here because the query itself retrieves only purposes having at least one version of those states).

But you're right, that's not clear at all. Gonna write comments 😁